modules/siddhi-core/src/main/java/org/wso2/siddhi/core/util/collection/operator/IncrementalAggregateCompileCondition.java

```diff
@@ -0,0 +1,266 @@
+                baseExecutor.cloneExecutor(key)).collect(Collectors.toList()));
+        newOutputExpressionExecutors.addAll(outputExpressionExecutors.stream().map(outputExpressionExecutor ->
+                outputExpressionExecutor.cloneExecutor(key)).collect(Collectors.toList()));
+        return new IncrementalAggregateCompileCondition(newTableCompiledConditions,
+                inMemoryStoreCompileCondition.cloneCompiledCondition(key),
+                finalCompiledCondition.cloneCompiledCondition(key), newBaseExecutors, aggregationTables,
+                inMemoryStoreMap, internalMetaStreamEvent, incrementalDurations, newOutputExpressionExecutors,
+                finalOutputMetaStreamEvent);
+    }
+
+    public StreamEvent find(StateEvent matchingEvent, TimePeriod.Duration perValue,
+                            boolean perEqualsRootExecutorDuration) {
+        ComplexEventChunk<StreamEvent> complexEventChunk = new ComplexEventChunk<>(true);
+        StreamEvent matchFromPersistedEvents = aggregationTables.get(perValue).find(matchingEvent,
+                tableCompiledConditions.get(perValue));
+        complexEventChunk.add(matchFromPersistedEvents);
+
+        findAndProcessInMemoryData(inMemoryStoreMap.get(perValue), matchingEvent, perValue);
+
+        if (!perEqualsRootExecutorDuration) {
+            // If the 'per' is for root executor, there would be no executors prior to that
+            for (int i = perValue.ordinal() - 1; i >= 0
+                    && incrementalDurations.contains(TimePeriod.Duration.values()[i]); i--) {
+                TimePeriod.Duration duration = TimePeriod.Duration.values()[i];
+                findAndProcessInMemoryData(inMemoryStoreMap.get(duration), matchingEvent, perValue);
+            }
+        }
+        for (Map<String, Object[]> executorsAndOutputDataMap : inMemoryAggregateMap.values()) {
+            for (Object[] executorsAndOutputData : executorsAndOutputDataMap.values()) {
+                List<ExpressionExecutor> executors = (List<ExpressionExecutor>) executorsAndOutputData[0];
+                Object[] outputData = (Object[]) executorsAndOutputData[1];
+                StreamEvent streamEvent = streamEventPoolForInternalMeta.borrowEvent();
+                streamEvent.setTimestamp((long) outputData[0]);
+                streamEvent.setOutputData(outputData);
+                complexEventChunk.add(streamEvent);
+                for (ExpressionExecutor expressionExecutor : executors) {
+                    expressionExecutor.execute(resetEvent);
+                }
+            }
+        }
+        return executeFinalCompileConditionOnWithinData(matchingEvent, complexEventChunk);
+    }
+
+    private void findAndProcessInMemoryData(Object inMemoryStore, StateEvent matchingEvent,
+                                            TimePeriod.Duration perValue) {
+        if (inMemoryStore instanceof ArrayList) {
+            for (int i = 0; i < ((ArrayList) inMemoryStore).size(); i++) {
```

Comment:
Same reason as for the first comment. In IncrementalExecutor, we keep all the running aggregates in one of four stores (baseIncrementalValueStore, baseIncrementalValueStoreMap, baseIncrementalValueStoreList or baseIncrementalValueGroupByStoreList). This method takes whatever the store that's passed to it, creates a complex event chunk out of the data in that store, and checks whether that data is a match using the find method of previously created inMemoryStoreCompileCondition. If it is a match, (that is it's within the given time period), we aggregate that data according to the time bucket it falls into.